File: test/storyarn_web/live/flow_live/player/components/player_outcome_test.exs
defmodule StoryarnWeb.FlowLive.Player.Components.PlayerOutcomeTest do
  use StoryarnWeb.ConnCase, async: true

  import Phoenix.LiveViewTest

  alias StoryarnWeb.FlowLive.Player.Components.PlayerOutcome

  defp build_slide(overrides \\ %{}) do
    Map.merge(
      %{
        label: "Game Over",
        outcome_color: "#ff0000",
        outcome_tags: ["bad ending", "death"],
        step_count: 12,
        choices_made: 5,
        variables_changed: 3
      },
      overrides
    )
  end

  defp build_workspace, do: %{slug: "my-workspace"}
  defp build_project, do: %{slug: "my-project"}
  defp build_flow, do: %{id: "flow-123"}

  # =============================================================================
  # player_outcome/1 rendering
  # =============================================================================

  describe "player_outcome/1 basic rendering" do
    test "renders outcome slide with title" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "player-slide-outcome"
      assert html =~ "Game Over"
    end

    test "renders outcome color accent bar" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{outcome_color: "#22c55e"}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "player-outcome-accent"
      assert html =~ "background-color: #22c55e"
    end

    test "renders without accent color when outcome_color is nil" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{outcome_color: nil}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "player-outcome-accent"
      refute html =~ "background-color:"
    end
  end

  describe "player_outcome/1 tags" do
    test "renders outcome tags" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{outcome_tags: ["victory", "heroic"]}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "player-outcome-tags"
      assert html =~ "victory"
      assert html =~ "heroic"
    end

    test "does not render tags section when outcome_tags is empty" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{outcome_tags: []}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      refute html =~ "player-outcome-tags"
    end
  end

  describe "player_outcome/1 stats" do
    test "renders step count stat" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{step_count: 42}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "player-outcome-stats"
      assert html =~ "footprints"
      assert html =~ "42"
    end

    test "renders choices made stat" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{choices_made: 7}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "mouse-pointer-click"
      assert html =~ "7"
    end

    test "renders variables changed stat" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{variables_changed: 15}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "15"
    end

    test "renders zero stats correctly" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(%{step_count: 0, choices_made: 0, variables_changed: 0}),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "player-outcome-stats"
    end
  end

  describe "player_outcome/1 actions" do
    test "renders restart button" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(),
          workspace: build_workspace(),
          project: build_project(),
          flow: build_flow()
        })

      assert html =~ "player-outcome-actions"
      assert html =~ "phx-click=\"restart\""
      assert html =~ "rotate-ccw"
    end

    test "renders back to editor link with correct path" do
      html =
        render_component(&PlayerOutcome.player_outcome/1, %{
          slide: build_slide(),
          workspace: %{slug: "ws-test"},
          project: %{slug: "proj-test"},
          flow: %{id: "flow-abc"}
        })

      assert html =~ "/workspaces/ws-test/projects/proj-test/flows/flow-abc"
      assert html =~ "arrow-left"
    end
  end
end

File: assets/app/test/modules/flows/player/PlayerOutcome.test.ts
import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import PlayerOutcome from "@modules/flows/player/PlayerOutcome.vue";
import type { OutcomeData } from "@modules/flows/player/PlayerOutcome.vue";

function buildSlide(overrides: Partial<OutcomeData> = {}): OutcomeData {
  return {
    type: "outcome" as const,
    label: "Game Over",
    outcome_color: "#ff0000",
    outcome_tags: ["bad ending", "death"],
    step_count: 12,
    choices_made: 5,
    variables_changed: 3,
    ...overrides,
  };
}

function mountOutcome(slide: OutcomeData = buildSlide(), editorUrl = "/flows/123") {
  return mount(PlayerOutcome, {
    props: { slide, editorUrl },
  });
}

describe("PlayerOutcome", () => {
  describe("basic rendering", () => {
    it("renders outcome with title", () => {
      const w = mountOutcome();

      expect(w.find(".player-slide-outcome").exists()).toBe(true);
      expect(w.find(".player-outcome-title").text()).toBe("Game Over");
    });

    it("renders accent bar with color", () => {
      const w = mountOutcome(buildSlide({ outcome_color: "#22c55e" }));
      const accent = w.find(".player-outcome-accent");
      expect(accent.attributes("style")).toMatch(
        /background-color:.*22c55e|background-color: rgb\(34, 197, 94\)/,
      );
    });

    it("renders accent bar without color when null", () => {
      const w = mountOutcome(buildSlide({ outcome_color: null }));
      const accent = w.find(".player-outcome-accent");
      expect(accent.attributes("style")).toBeUndefined();
    });
  });

  describe("tags", () => {
    it("renders outcome tags as badges", () => {
      const w = mountOutcome(buildSlide({ outcome_tags: ["victory", "heroic"] }));

      expect(w.find(".player-outcome-tags").exists()).toBe(true);
      const badges = w.findAll("[data-slot='badge']");
      expect(badges).toHaveLength(2);
      expect(w.text()).toContain("victory");
      expect(w.text()).toContain("heroic");
    });

    it("hides tags section when empty", () => {
      const w = mountOutcome(buildSlide({ outcome_tags: [] }));
      expect(w.find(".player-outcome-tags").exists()).toBe(false);
    });
  });

  describe("stats", () => {
    it("renders step count", () => {
      const w = mountOutcome(buildSlide({ step_count: 42 }));
      expect(w.text()).toContain("42");
    });

    it("renders choices made", () => {
      const w = mountOutcome(buildSlide({ choices_made: 7 }));
      expect(w.text()).toContain("7");
    });

    it("renders variables changed", () => {
      const w = mountOutcome(buildSlide({ variables_changed: 15 }));
      expect(w.text()).toContain("15");
    });

    it("renders zero stats", () => {
      const w = mountOutcome(buildSlide({ step_count: 0, choices_made: 0, variables_changed: 0 }));
      expect(w.find(".player-outcome-stats").exists()).toBe(true);
    });
  });

  describe("actions", () => {
    it("emits restart on play again click", async () => {
      const w = mountOutcome();
      const btns = w.findAll("[data-slot='button']");
      const playAgain = btns.find((b) => b.text().includes("Play again"))!;
      await playAgain.trigger("click");
      expect(w.emitted("restart")).toHaveLength(1);
    });

    it("renders back to editor link with correct path", () => {
      const w = mountOutcome(buildSlide(), "/workspaces/ws/projects/proj/flows/abc");
      const link = w.find("a");
      expect(link.attributes("href")).toBe("/workspaces/ws/projects/proj/flows/abc");
    });
  });
});
